### ISSUE TITLE
Client ID scope is replaced with slash

### ISSUE THREAD

**Issue by baSSiLL** [AzureAD/microsoft-authentication-library-for-js#2451]
>> Please follow the issue template below. Failure to do so will result in a delay in answering your question.

## Library
- [ ] `msal@1.x.x` or `@azure/msal@1.x.x`
- [x] `@azure/msal-browser@2.3.1`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

**Important**: Please fill in your exact version number above, e.g. `msal@1.1.3`.

## Framework
AAD B2C with Vue JS app.

## Description
I want to authenticate calls to the backend API of an SPA app. For simplicity the frontend SPA app and the backend app use the same app registration within AD B2C. This allows to avoid creating an explicit scope for the API and simply use a client ID as a scope. However, when such scope is specified in calls to `loginRedirect` or `acquireTokenSilent` it's replaced with "/" in a response object.
```
{
  accessToken: "...",
  ...
  scopes: ["/"],
  ...
}
```
And a token is stored in cache with this "slash" scope. So when I call `acquireTokenSilent` next time with client ID in scopes it misses the cache. And if I try to specify "/" in scopes, it hits the cache. But `loginRedirect` doesn't accept "/" in scopes returning an error message that the requested scope is invalid.
As a workaround, I resorted to creating an explicit scope in the app registration and granting permission to that scope to the app itself. Specifying this explicit scope URI in scopes works as expected.

## Security
- [ ] Is this issue security related?

## Regression
- [ ] Did this behavior work before? 
No

## MSAL Configuration

```js
{
  auth: {
    clientId: 'CLIENT_ID',
    authority: 'https://TENANT.b2clogin.com/TENANT.onmicrosoft.com/B2C_1_signin',
    knownAuthorities: ['TENANT.b2clogin.com'],
    redirectUri: 'http://localhost:8080/auth',
    postLogoutRedirectUri: 'http://localhost:8080/login',
    navigateToLoginRequestUrl: false
  },
  cache: {
    cacheLocation: 'localStorage',
    storeAuthStateInCookie: true
  }
}
```

## Reproduction steps
```js
const instance = new Msal.PublicClientApplication(config)
// Sign in with redirect
instance.loginRedirect({
  scopes: ['CLIENT_ID']
})
// After successful sign in
const response1 = await instance.handleRedirectPromise()
// Access token is obtained
// response.fromCache == false
// response.scopes == ['/']
const account = instance.getAllAccounts()[0]
const response2 = await instance.acquireTokenSilent({
  account,
  scopes: ['CLIENT_ID']
})
// Access token is obtained
// response.fromCache == false (expect to be true)
// response.scopes == ['/']
```

## Expected behavior
MSAL doesn't change the scope value. Otherwise it supports "/" instead of "client ID" scope.

## Browsers/Environment
- [x] Chrome
- [ ] Firefox
- [x] Edge
- [ ] Safari
- [ ] IE
- [ ] Other (Please add browser name here)

**Comment by jasonnutter** [comment]
@baSSiLL Thanks for the information. We're checking with the B2C service about this behavior, and will follow up when we know more.

cc: @ktsakas

**Comment by ktsakas** [comment]
B2C does not allow sending an empty scope, at least "openid" needs to be passed as a scope. Could you take a look at the actual request and see what the scope is there? If you could provide a fiddler trace that would be ideal.

**Comment by jasonnutter** [comment]
@ktsakas The app is sending their client id as the scope value (and then MSAL.js would add openid and profile to that).

**Comment by baSSiLL** [comment]
Pinging in a hope for resolution. @jasonnutter any news here?

**Comment by sameerag** [comment]
We are working with B2C Service team to help us understand why `ClientId` as a scope is being replaced by a `/`. We haven't yet heard a definitive answer. Will update here as soon as we have the info.

**Comment by vinizinmoraes** [comment]
Hi,

Any news about this bug or a workaround? I am trying to update from msal 1.x to 2.x and just realize right now that everything is correct but the scope is been change from client id to "/". Thanks!

**Comment by baSSiLL** [comment]
I outlined the workaround which worked for me in the original post:
> As a workaround, I resorted to creating an explicit scope in the app registration and granting permission to that scope to the app itself. Specifying this explicit scope URI in scopes works as expected.

**Comment by vinizinmoraes** [comment]
@baSSiLL Can you show me how to create this explicit scope URI? I asked it to me collegue who has access to the azure and he doesnt know how to do it. Can you help me out ?

**Comment by baSSiLL** [comment]
@vinizinmoraes Check out this post
https://codeburst.io/practical-azure-secure-a-net-core-web-api-using-azure-ad-b2c-fde234a8e819#b8ca
They describe the process of creating a web API scope and granting access to it to a client web app, with screenshots of Azure Portal. In your case a single AD app serves both as a web API and its client, so you do everything with a single app registration ("Expose and API" and "API permissions" blades).

**Comment by vinizinmoraes** [comment]
@baSSiLL thanks!

**Comment by pkanher617** [comment]
@baSSiLL @vinizinmoraes Were you able to figure out how to work around this issue?

**Comment by baSSiLL** [comment]
@pkanher617 @sameerag So you just want to close the issue without a fix? Any update from AD B2C team?

> We are working with B2C Service team to help us understand why ClientId as a scope is being replaced by a /. We haven't yet heard a definitive answer. Will update here as soon as we have the info.

**Comment by pkanher617** [comment]
@baSSiLL We are still waiting on an update from the B2C team. However from MSAL.js perspective there is nothing further we can do until they push out an update for this. I can leave this open but I suggest going to the [B2C support page](https://docs.microsoft.com/en-us/azure/active-directory-b2c/support-options) and filing a ticket there to track it from the B2C service's end. I'll also make sure to add this workaround to the FAQ.

**Comment by Azzmurr** [comment]
Also faced with this issue

**Comment by Azzmurr** [comment]
@pkanher617 Also faced with this issue. Any updates?

**Comment by vinizinmoraes** [comment]
> @baSSiLL @vinizinmoraes Were you able to figure out how to work around this issue? Can I close this or are there further issues happening?

I used the @baSSiLL workaround. But its really strange, on 1.4.0 I was using the clientId as scope without any problems.

**Comment by tnorling** [comment]
> I used the @baSSiLL workaround. But its really strange, on 1.4.0 I was using the clientId as scope without any problems.

@vinizinmoraes 1.4.0 is a completely different library that uses an entirely different authentication flow. The root cause of this issue is that the /token endpoint is returning the `/` instead of `clientId` whereas in version 1.x tokens are retrieved from the /authorize endpoint which is correctly returning the `clientId`.  

> Also faced with this issue. Any updates?

@Azzmurr The B2C service team has committed to this work but we cannot provide timelines on behalf of other teams. If you'd like to track this with them directly you can open a support ticket at the link mentioned a few comments above and in the meantime we'll leave this open until we've verified the fix has been rolled out.

**Comment by jacknewberry** [comment]
Could you consider adding some more comprehensive documentation about scopes to the docs for this library.

There seem to be some subtleties in how tokens are retrieved and cached, what scopes are used by default, and how the Azure AD responds.
At least one of these is the cause of this issue being opened.

Some issues include:
- several scopes appear to be added by default to every request (`"openid"`, `"profile"`, `"offline_access"`)
- if only these scopes are sent, the`acquireTokenSilent` call succeeds but no access token is returned
  - the application registration ID can be passed as a scope to cause an access token to be returned
- the docs are not clear that scopes are used in the key when caching tokens (though this is intuitive with a little consideration)
- there is an issue where Azure AD converts the scope `<app_id>` into `"/"`, which causes caching to fail
  - this can be worked around by creating a dummy scope to use instead of the app registration ID
 
I appreciate that some of these issues are due to interactions with other components and authentication standards. Nonetheless they are recurrent issues that users of your library are encountering so it would be great if you could help to address them.

This library is the first point of contact for web developers trying to use Azure AD.

**Comment by tnorling** [comment]
@jacknewberry Thanks for the feedback! We certainly understand documentation can always be improved. In this case you may find the following docs of interest:

- [All about scopes](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/resources-and-scopes.md)
- [Working with B2C](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/working-with-b2c.md)
- [FAQ mentioning default scopes](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/FAQ.md#im-seeing-scopes-openid-profile-email-offline_access-and-userread-in-my-tokens-even-though-i-havent-requested-them-what-are-they)

As this issue is long running and I don't have a timeline for resolution yet I've also opened a PR to add a specific note about this issue to our docs. See #3287. Generally speaking, I would not expect bugs to be added to our docs so searching through our open issues is still a good idea when you have questions or if something isn't quite working the way you expect.

**Comment by mindingdata** [comment]
>Generally speaking, I would not expect bugs to be added to our docs so searching through our open issues is still a good idea when you have questions or if something isn't quite working the way you expect.

Tend to agree but if a "bug" has been ongoing for the past 9 months. Not sure at what point this turns into a feature. 

I came here searching for this as we moved from MSAL 1.X to 2. It is very frustrating that such a common scenario now doesn't work correctly.

**Comment by sameerag** [comment]
This is a dependency on the B2C service and we have been trying our best to follow through. They are unable to fix this right away as this isn't straight forward for them. I will amplify this again in the right channels. Apologies for the frustrating wait.

**Comment by tnorling** [comment]
I've just confirmed with the B2C service team that a fix for this has been rolled out. Thanks everyone for your patience! Closing this but please feel free to open a new issue if you continue to have problems.

**Comment by cranarma** [comment]
I'm still experiencing this issue. The "scope" value in the response from B2C is still a slash.
